timeseries/src/autogluon/timeseries/models/autogluon_tabular/tabular_model.py

```diff
@@ -163,78 +140,130 @@ def apply_mask(
-            # This mimics what will happen at test time, when we simultaneously predict the next `prediction_length` values
-            num_windows = (len(time_series) - 1) // self.prediction_length
-            # We don't hide any past values for the first `remainder` values, otherwise the features will be all empty
-            remainder = len(time_series) - num_windows * self.prediction_length
-            num_hidden = np.concatenate([np.zeros(remainder), np.tile(np.arange(self.prediction_length), num_windows)])
-            target_lags = apply_mask(target_lags, num_hidden=num_hidden, lag_indices=self._target_lag_indices)
-            feature_dfs = [target_lags, time_series[[self.target]]]
-
-            if self.metadata.past_covariates_real:
-                past_covariates_lags = get_lags(
-                    time_series[self.metadata.past_covariates_real], self._past_covariates_lag_indices
-                )
-                past_covariates_lags = apply_mask(
-                    past_covariates_lags,
-                    num_hidden=num_hidden,
-                    lag_indices=self._past_covariates_lag_indices,
-                    num_columns=len(self.metadata.past_covariates_real),
-                )
-                feature_dfs.append(past_covariates_lags)
+            num_rows = min(max_rows_per_item, len(ts))
+            features = np.full([num_rows, len(lag_indices)], fill_value=np.nan)
+            for i in range(1, num_rows + 1):
+                target_idx = len(ts) - i
+                selected_lags = lag_indices[lag_indices <= target_idx]
+                features[num_rows - i, np.arange(len(selected_lags))] = ts[target_idx - selected_lags]
+            if mask:
+                num_windows = (len(ts) - 1) // prediction_length
+                # We don't hide any past values for the first `remainder` values, otherwise the features will be all empty
+                remainder = len(ts) - num_windows * prediction_length
+                num_hidden = np.concatenate([np.zeros(remainder), np.tile(np.arange(prediction_length), num_windows)])
+                features = apply_mask(features, num_hidden[-num_rows:], lag_indices)
+            return features
 
-            if self.metadata.known_covariates_real:
-                known_covariates_lags = get_lags(
-                    time_series[self.metadata.known_covariates_real], self._known_covariates_lag_indices
+        def get_lag_features(
+            all_series: List[np.ndarray],
+            lag_indices: np.ndarray,
+            prediction_length: int,
+            max_rows_per_item: int,
+            mask: bool,
+            name: str,
+        ):
+            """Generate lag features for all time series in the dataset.
+
+            See the docstring of get_lags for the description of the parameters.
+            """
+            lags_per_item = Parallel(n_jobs=-1)(
```

Comment:
[non blocker] we should also allow users to override the number of workers for the processing job. Something similar we exposed via AutoMM at [here](https://auto.gluon.ai/stable/tutorials/multimodal/advanced_topics/customization.html#env-num-workers). If not overridden , we can use -1

Reply:
Totally agree! This is something that is currently not handled uniformly in the entire timeseries module - there are quite a few places where the number of used CPUs / GPUs is either chosen automatically or hardcoded. Fixing this is on the roadmap for v0.8.

How about we keep this as-is for now and get back to it when aligning all the resource allocation logic with Tabular/AutoMM as a separate PR?

Reply:
Yes, let's create an issue and track it for 0.8 release